pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,269 @@
+	status := &statefulSet.Status
+	if status.CurrentReplicas == 0 || status.CurrentReplicas < status.Replicas {
+		log.Info(" resize pvc statefulset does not have all replicas up")
+		return NotReadyErr{Err: errors.New("resize pvc statefulset does not have all replicas up")}
+	}
+
+	// Maybe this should be an error since we should not have this, but I wanted to check anyways
+	if len(statefulSet.Spec.VolumeClaimTemplates) == 0 {
+		log.Info("Skipping PVC resize as PVCs do not exist")
+		return nil
+	}
+
+	clientset, err := kubernetes.NewForConfig(rp.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create kubernetes clientset")
+	}
+
+	stsStorageSizeDeployed := statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()
+	stsStorageSizeSet := cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage()
+
+	// If the sizes match do not resize
+	if stsStorageSizeDeployed.Equal(stsStorageSizeSet.DeepCopy()) {
+		log.Info("Skipping PVC resize as sizes match")
+		return nil
+	}
+
+	log.Info("Starting PVC resize")
+
+	// Find all of the PVCs and resize them
+	if err := rp.findAndResizePVC(ctx, statefulSet, cluster, clientset); err != nil {
+		return errors.Wrapf(err, "updating PVCs for statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+	}
+
+	log.Info("Starting updating sts")
+
+	// Update the STS with the correct volume size, in case more pods are created
+	// We will create a copy and update the copy, and then delete the original without
+	// deleting the Pods.  The new sts is then used to create a new statefulset.
+	if err := rp.updateSts(ctx, statefulSet, cluster); err != nil {
+		return errors.Wrapf(err, "updating statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+	}
+
+	if !cluster.Spec().DataStore.SupportsAutoResize {
+		log.Info("Starting rolling sts")
+		// Roll the entire STS in order for the Pods to resize
+		if err := rp.rollSts(ctx, cluster, clientset); err != nil {
+			return errors.Wrapf(err, "error restarting statefulset %s.%s", cluster.Namespace(), cluster.StatefulSetName())
+		}
+	} else {
+		log.Info("Volumes support autoresizing so not restarting STS Pods")
+	}
+
+	log.Info("PVC resize completed")
+	CancelLoop(ctx)
+
+	return nil
+}
+
+// updateSts updates the size of an STS' VolumeClaimTemplate to match the new size in the CR.
+// In order to update the volume claim template we have to delete the STS without cascading and then
+// create the sts.
+func (rp *resizePVC) updateSts(ctx context.Context, sts *appsv1.StatefulSet, cluster *resource.Cluster) error {
+
+	// delete the original sts, but do not delete the Pods
+	orphan := metav1.DeletePropagationOrphan
+	if err := rp.client.Delete(ctx, sts, &client.DeleteOptions{PropagationPolicy: &orphan}); err != nil {
+		return err
+	}
+
+	f := func() error {
+		return rp.recreateSTS(ctx, cluster)
+	}
+
+	b := backoffFactory(5 * time.Minute)
+	return backoff.Retry(f, backoff.WithContext(b, ctx))
```

Comment:
@chrisseto, here is my backoff retry for creating the new STS.  Once K8s has reconciled the deletion of the sts it creates the updated version.